### PR TITLE
Normalize files

### PR DESCRIPTION
**Pull request by tbeadle**
Insert the detonation-independent fields for files in to a separate collection and reference it from the analysis document.

Many of the fields stored for file objects with the same hash (target.file, dropped, CAPE.payloads, and procdump) are the same, regardless of the detonation. These can take a significant amount of space in the mongo database, so store them in a separate collection holding just those attributes once for each unique file. Keep track of the task ID's that reference them, so that they can be cleaned up later.

### CODE REVIEW

**Comment by doomedraven** [comment]
nice stuff

**Comment by stamparm** [comment]
@tbeadle @doomedraven and how to "clean them later"? what's the new workflow?

AFAICS, currently the `/apiv2/tasks/delete/<task_id>/` has changed the behavior (i.e. it is broken compared to before) as those extra collections are being filled (e.g. `files`), even though the tasks are explicitly deleted

TL;DR: before, `delete/<task_id>` deleted everything, now junk is pilled up in these separate collections

**Comment by doomedraven** [comment]
yes i spot yesterday more broken stuff like find_sample_by_hash in database

**Comment by tbeadle** [comment]
@stamparm 

> @tbeadle @doomedraven and how to "clean them later"? what's the new workflow?
> 
> AFAICS, currently the `/apiv2/tasks/delete/<task_id>/` has changed the behavior (i.e. it is broken compared to before) as those extra collections are being filled (e.g. `files`), even though the tasks are explicitly deleted
> 
> TL;DR: before, `delete/<task_id>` deleted everything, now junk is pilled up in these separate collections

From https://github.com/kevoreilly/CAPEv2/blob/master/changelog.md#1992023:

```
It is recommended to add a regular cron job to call cd /opt/CAPEv2 && sudo -u cape poetry run python ./utils/cleaners.py --delete-unused-file-data-in-mongo to prune these entries that are no longer needed.
```

We track which tasks have references to which file objects using the `_task_ids` key of the `files` collection. Using `cleaners.py --delete-unused-file-data-in-mongo` as described above will find ones that have no tasks that reference them any more and delete them. This is done to optimize for deleting large ranges of tasks, such as with `cleaners.py -drs ### -dre ###`. Since Mongo, AFAIK, does not have a way to update the `_task_ids` key and return its new value in the same call (like `UPDATE .... RETURNING ...` in SQL), we'd have to do a separate query after each file being deleted to see if its `_task_ids` key is empty. This could significantly slow down that operation of cleaning many tasks.

Using this asynchronous operation of cleaning those unused file objects is able to be done in 1 mongodb call that should be relatively quick, given that an index for `_task_ids` is created in startup.py.

@doomedraven can you explain more about the find_sample issue you're seeing?

**Comment by stamparm** [comment]
Can I ask why there was a need for such drastic changes? Everything was
working fine, and now this seems to me like an overengineering (i.e. too
much complications for little to no gain IMHO)

**Comment by doomedraven** [comment]
> @stamparm
> 
> > @tbeadle @doomedraven and how to "clean them later"? what's the new workflow?
> > AFAICS, currently the `/apiv2/tasks/delete/<task_id>/` has changed the behavior (i.e. it is broken compared to before) as those extra collections are being filled (e.g. `files`), even though the tasks are explicitly deleted
> > TL;DR: before, `delete/<task_id>` deleted everything, now junk is pilled up in these separate collections
> 
> From https://github.com/kevoreilly/CAPEv2/blob/master/changelog.md#1992023:
> 
> ```
> It is recommended to add a regular cron job to call cd /opt/CAPEv2 && sudo -u cape poetry run python ./utils/cleaners.py --delete-unused-file-data-in-mongo to prune these entries that are no longer needed.
> ```
> 
> We track which tasks have references to which file objects using the `_task_ids` key of the `files` collection. Using `cleaners.py --delete-unused-file-data-in-mongo` as described above will find ones that have no tasks that reference them any more and delete them. This is done to optimize for deleting large ranges of tasks, such as with `cleaners.py -drs ### -dre ###`. Since Mongo, AFAIK, does not have a way to update the `_task_ids` key and return its new value in the same call (like `UPDATE .... RETURNING ...` in SQL), we'd have to do a separate query after each file being deleted to see if its `_task_ids` key is empty. This could significantly slow down that operation of cleaning many tasks.
> 
> Using this asynchronous operation of cleaning those unused file objects is able to be done in 1 mongodb call that should be relatively quick, given that an index for `_task_ids` is created in startup.py.
> 
> @doomedraven can you explain more about the find_sample issue you're seeing?

it just was checking directly instead of add ref_file, my idea is to replace it with just search function from web_utils.py but i don't have free time those days to do that properly

**Comment by tbeadle** [comment]
> Can I ask why there was a need for such drastic changes? Everything was working fine, and now this seems to me like an overengineering (i.e. too much complications for little to no gain IMHO)

We were regularly having detonation documents that would fail to be inserted into Mongo because they exceeded Mongo's document size limit. This was a way to reduce not only the document size but also the overall Mongo storage due to the deduplication.

**Comment by stamparm** [comment]
@tbeadle you sure about "deduplication"? do you have some numbers on reduction, as to be honest, I have doubts on such statement (i.e. I expect that vast majority of files stuffed into mongodb from CAPE is fairly unique)

**Comment by doomedraven** [comment]
files are uniq if you don't use a lot of `file_extra_info` features(selfextract.conf and custom modules), we do use it a lot at my $dayjob and i saw tons of duplicated files coming from different "parents"

**Comment by doomedraven** [comment]
btw speaking of duplicated, we got rid of them in mongo, but we didn't get rid of them on storage. would be nice to develop solution for that too, i will ping you later tbeadle to discuss some ideas
